src/librustc_mir_build/hair/constant.rs

```diff
@@ -20,50 +20,35 @@ crate fn lit_to_const<'tcx>(
         Ok(ConstValue::Scalar(Scalar::from_uint(result, width)))
     };
 
-    let lit = match *lit {
-        ast::LitKind::Str(ref s, _) => {
+    let lit = match (lit, &ty.kind) {
+        (ast::LitKind::Str(s, _), ty::Ref(_, TyS { kind: ty::Str, .. }, _)) => {
             let s = s.as_str();
             let allocation = Allocation::from_byte_aligned_bytes(s.as_bytes());
             let allocation = tcx.intern_const_alloc(allocation);
             ConstValue::Slice { data: allocation, start: 0, end: s.len() }
         }
-        ast::LitKind::ByteStr(ref data) => {
-            if let ty::Ref(_, ref_ty, _) = ty.kind {
-                match ref_ty.kind {
-                    ty::Slice(_) => {
-                        let allocation = Allocation::from_byte_aligned_bytes(data as &Vec<u8>);
-                        let allocation = tcx.intern_const_alloc(allocation);
-                        ConstValue::Slice { data: allocation, start: 0, end: data.len() }
-                    }
-                    ty::Array(_, _) => {
-                        let id = tcx.allocate_bytes(data);
-                        ConstValue::Scalar(Scalar::Ptr(id.into()))
-                    }
-                    _ => {
-                        bug!("bytestring should have type of either &[u8] or &[u8; _], not {}", ty)
-                    }
-                }
-            } else {
-                bug!("bytestring should have type of either &[u8] or &[u8; _], not {}", ty)
-            }
+        (ast::LitKind::ByteStr(data), ty::Ref(_, TyS { kind: ty::Slice(_), .. }, _)) => {
+            let allocation = Allocation::from_byte_aligned_bytes(data as &Vec<u8>);
+            let allocation = tcx.intern_const_alloc(allocation);
+            ConstValue::Slice { data: allocation, start: 0, end: data.len() }
```

Comment:
Reminder to self that reviewing this is best done in "No Whitespace" mode.

Reply:
Indeed :)